presto-benchmark-runner/src/main/java/com/facebook/presto/benchmark/framework/ConcurrentPhaseExecutor.java

```diff
@@ -0,0 +1,135 @@
+        log.info("Starting concurrent phase '%s' with max concurrency %s", phase.getName(), maxConcurrency);
+
+        ExecutorService executor = newFixedThreadPool(maxConcurrency);
+
+        try {
+            CompletionService<BenchmarkQueryEvent> completionService = new ExecutorCompletionService<>(executor);
+
+            for (String queryName : phase.getQueries()) {
+                BenchmarkQuery benchmarkQuery = overrideSessionProperties(suite.getQueries().get(queryName), suite.getSessionProperties());
+                completionService.submit(() -> runQuery(benchmarkQuery));
+            }
+
+            return reportProgressUntilFinished(phase, completionService);
+        }
+        finally {
+            executor.shutdownNow();
+        }
+    }
+
+    private BenchmarkPhaseEvent reportProgressUntilFinished(
+            ConcurrentExecutionPhase phase,
+            CompletionService<BenchmarkQueryEvent> completionService)
+    {
+        String phaseName = phase.getName();
+        int completed = 0;
+        double lastProgress = 0;
+        int queriesSubmitted = phase.getQueries().size();
+        Map<Status, Integer> statusCount = new EnumMap<>(Status.class);
+
+        while (completed < queriesSubmitted) {
+            try {
+                BenchmarkQueryEvent event = completionService.take().get();
+                postEvent(event);
+                completed++;
+                statusCount.compute(event.getEventStatus(), (status, count) -> count == null ? 1 : count + 1);
+                if (event.getEventStatus() == FAILED && !continueOnFailure) {
+                    return BenchmarkPhaseEvent.failed(phaseName, event.getErrorMessage());
+                }
+
+                double progress = ((double) completed) / queriesSubmitted * 100;
+                if (progress - lastProgress > 0.5 || completed == queriesSubmitted) {
+                    log.info("Progress: %s succeeded, %s failed, %s submitted, %.2f%% done",
+                            statusCount.getOrDefault(SUCCEEDED, 0),
+                            statusCount.getOrDefault(FAILED, 0),
+                            queriesSubmitted,
+                            progress);
+                    lastProgress = progress;
+                }
+            }
+            catch (InterruptedException e) {
+                currentThread().interrupt();
+                if (!continueOnFailure) {
+                    return BenchmarkPhaseEvent.failed(phaseName, e.toString());
+                }
+            }
+            catch (ExecutionException e) {
+                if (!continueOnFailure) {
+                    return BenchmarkPhaseEvent.failed(phaseName, e.toString());
+                }
+            }
+        }
+        if (statusCount.getOrDefault(FAILED, 0) > 0) {
+            return BenchmarkPhaseEvent.completedWithFailures(phaseName, format("%s out of %s submitted queries failed", statusCount.get(FAILED), queriesSubmitted));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: queriesSubmitted (declared at line 94). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 130. Consider formatting this data with &#39;%d&#39; instead.